include/material/mohr_coulomb.tcc

```diff
@@ -0,0 +1,576 @@
+      this->compute_yield(epsilon, rho, theta);
+  int yield_type = this->check_yield(yield_function, epsilon, rho, theta);
+  if (yield_type != 1 && (epds - peak_epds_) > 0. && (crit_epds_ - epds) > 0.) {
+    phi_ = phi_min + ((phi_max - phi_min) * (epds - crit_epds_) /
+                      (peak_epds_ - crit_epds_));
+    psi_ = psi_min + ((psi_max - psi_min) * (epds - crit_epds_) /
+                      (peak_epds_ - crit_epds_));
+    c_ = c_min +
+         ((c_max - c_min) * (epds - crit_epds_) / (peak_epds_ - crit_epds_));
+  } else if (yield_type != 1 && (epds - crit_epds_) >= 0.) {
+    phi_ = phi_min;
+    psi_ = psi_min;
+    c_ = c_min;
+  }
+
+  // Yield function for the current stress state
+  yield_function = this->compute_yield(epsilon, rho, theta);
+  yield_type = this->check_yield(yield_function, epsilon, rho, theta);
+  double softening = 0.;
+
+  // Compute plastic multiplier from the current stress state
+  Vector6d df_dsigma = Vector6d::Zero();
+  Vector6d dp_dsigma = Vector6d::Zero();
+  this->compute_df_dp(yield_type, j2, j3, rho, theta, stress, epds, &df_dsigma,
+                      &dp_dsigma, &softening, ptr);
+  // Check the epds
+  if (epds_final < peak_epds_ && epds > peak_epds_) softening = 0;
+  double lambda = df_dsigma.dot(this->de_ * dstrain) /
+                  ((df_dsigma.dot(this->de_ * dp_dsigma)) + softening);
+
+  // Compute the trial stress
+  j2 = j3 = rho = theta = 0.;
+  this->compute_rho_theta(trial_stress, &j2, &j3, &rho, &theta);
+
+  epsilon =
+      (1. / sqrt(3.)) * (trial_stress(0) + trial_stress(1) + trial_stress(2));
+
+  // Trial yield function
+  Eigen::Matrix<double, 2, 1> yield_function_trial =
+      this->compute_yield(epsilon, rho, theta);
+  int yield_type_trial =
+      this->check_yield(yield_function_trial, epsilon, rho, theta);
+
+  double softening_trial = 0.;
+  Vector6d df_dsigma_trial = Vector6d::Zero();
+  Vector6d dp_dsigma_trial = Vector6d::Zero();
+  this->compute_df_dp(yield_type_trial, j2, j3, rho, theta, trial_stress, epds,
+                      &df_dsigma_trial, &dp_dsigma_trial, &softening_trial,
+                      ptr);
+
+  // Check the epds
+  if (epds_final < peak_epds_ && epds > peak_epds_) softening_trial = 0;
+  double lambda_trial = 0.;
+
+  if (yield_type_trial == 1)
+    lambda_trial =
+        yield_function_trial(0) /
+        ((df_dsigma_trial.transpose() * de_).dot(dp_dsigma_trial.transpose()) +
+         softening_trial);
+  else
+    lambda_trial =
+        yield_function_trial(1) /
+        ((df_dsigma_trial.transpose() * de_).dot(dp_dsigma_trial.transpose()) +
+         softening_trial);
+  // Compute the correction stress
+  double p_multiplier = 0.;
+  Vector6d dp_dsigma_final = Vector6d::Zero();
+  bool update_pds = false;
+  // check if it is a load process or an unload process
+  double dyfun_t = yield_function_trial(0) - yield_function(0);
+  double dyfun_s = yield_function_trial(1) - yield_function(1);
+  if (yield_type != 0) {
```

Comment:
Needs clarification

Reply:
Let's keep this `if` statement for now. And it is to assign the correct plastic multiplier.